### PR TITLE
Once java always java

### LINKED ISSUE #2163: removing and adding java destination via reload cause crash
# syslog-ng

## Version of syslog-ng

Any version that supports java destination.

## Platform

Any platform where java destination is present. (But only checked on linux.)

# Issue

## Steps to reproduce

1. configure any java destination
2. start syslog-ng
3. remove java destination from configuration
4. reload syslog-ng
5. change the configuration again to include a java destination
6. reload syslog-ng
7. crash

### PR DESCRIPTION
**Pull request by Kokan**
Once a java is configured, and the JVM is loaded, and it is not possible to safely unload the JVM.
Because there cannot be two or more JVM in one process, the second time syslog-ng tries to create a JVM it wont succeeed.

The plan is to keep the JVM once created as long as syslog-ng is runnig.

Fixes #2163

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

File: modules/java/native/java_machine.c [hunk 1]
```diff
@@ -113,6 +113,12 @@ void
 java_machine_unref(JavaVMSingleton *self)
 {
   g_assert(self == global_jvm);
+  /* The last reference is always hold by a AH_SHUTDOWN app hook, when there is no java configured.
+   * The only way to get ride of it to shutdown syslog-ng.  */
```
Comment (MrAnno) [inline modules/java/native/java_machine.c]: Typo: get rid of

File: modules/java/native/java_machine.c [hunk 2]
```diff
@@ -113,6 +113,12 @@ void
...
+  if (g_atomic_counter_racy_get(&self->ref_cnt) == 1)
```
Comment (MrAnno) [inline modules/java/native/java_machine.c]: I wouldn't access an atomic counter in a racy way even if atomicity is guaranteed here.
We had difficulties to trace such things after the code changed a few years later.

I would also remove `g_atomic_counter_racy_get` completely.

File: modules/java/native/java_machine.c [hunk 3]
```diff
@@ -113,6 +113,12 @@ void
...
+    {
+      msg_warning("There is no need for the JVM, but it is not stoped, if you want to stop it, you have to restart syslog-ng");
```
Comment (MrAnno) [inline modules/java/native/java_machine.c]: I'd rephrase this sentence to make it more understandable, for example:
The JVM instance has become unused, but it has not been stopped. Restart syslog-ng, if you want to stop it.

A side note: this warning message will be displayed every time you stop syslog-ng (not just on reloads), if I understand it correctly.

Comment (MrAnno) [reply modules/java/native/java_machine.c]: (There is a JNI method called `DestroyJavaVM`, but it seems it is not guaranteed that it will stop the VM.)

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by Kokan** [comment]
@kira-syslogng retest this please

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by Kokan** [comment]
Waiting for testing: @mitzkia

**Comment by mitzkia** [comment]
Sorry for the delay. 
I have started to test the fix, and found a SIGABRT during the test.
It needs more investigation to understand it caused by the fix or not.

**Comment by lbudai** [comment]
@mizkia, @kokan: I don’t know all the details but if it’s not possible to test these changes until another bug not fixed, then open an issue for this bug, make this pr blocking on that issue. 
And provide enough information on the bug to make it possible developers to fix it.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by mitzkia** [comment]
@kira-syslogng retest this please test branch=micek_java_destination_crash;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by mitzkia** [comment]
@kira-syslogng retest this please test branch=micek_java_destination_crash;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by mitzkia** [comment]
@kira-syslogng retest this please test branch=micek_java_destination_crash;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by mitzkia** [comment]
@kira-syslogng retest this please test branch=micek_java_destination_crash;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by mitzkia** [comment]
Automated test is created, test is passed

**Comment by mitzkia** [comment]
@kira-syslogng retest this please
